File: frontend/src/app/modules/shared/components/ui/search-result.component/services/search.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { map } from 'rxjs';
import { API_URL } from '../../../../../../utils/constants';
import { groupSchedulesByDate } from '../../../../../../utils/scheduleMapper';
import { Hospital } from '../components/hospital-card.component/hospital-card.component';
import { Doctor } from '../search-result.model';

export enum SearchType {
  INSTITUTION = 'institution',
  DOCTOR = 'doctor',
}

export interface SearchResponse {
  result: Doctor[] | Hospital[];
}

export interface SearchQuery {
  query: string;
  category: string;
  location?: string;
  specialization?: string;
}

export interface ApiDoctor {
  name: string;
  numOfRatings: number;
  rating: number;
  surname: string;
  schedules: {
    startTime: string;
    isBooked: boolean;
    id: string;
  }[];
  specialisations: string[];
  addresses: {
    first: {
      institutionId: string;
      institutionName: string;
    };
    second: string;
  }[];
  image: string;
  id: string;
}

@Injectable({
  providedIn: 'root',
})
export class SearchService {
  private readonly http = inject(HttpClient);

  public search(searchQuery: SearchQuery) {
    let params = new HttpParams();

    params = params.set('type', searchQuery.category);

    if (searchQuery.location) {
      params = params.set('city', searchQuery.location);
    }

    if (searchQuery.specialization) {
      params = params.set('specialisations', searchQuery.specialization);
    }

    const query = searchQuery.query || '';
    return this.http
      .get<SearchResponse>(`${API_URL}/search/${query}`, {
        params,
      })
      .pipe(
        map((response) => {
          if (searchQuery.category === SearchType.DOCTOR) {
            const doctors = (response.result as unknown[]).map((doc) =>
              this.mapApiDoctorToComponentFormat(doc as ApiDoctor),
            );
            return { result: doctors } as SearchResponse;
          }
          return response;
        }),
      );
  }

  public mapApiDoctorToComponentFormat(apiDoctor: ApiDoctor): Doctor {
    return {
      id: apiDoctor.id,
      name: `${apiDoctor.name} ${apiDoctor.surname}`,
      specialisation: apiDoctor.specialisations.join(', '),
      rating: apiDoctor.rating,
      reviewsCount: apiDoctor.numOfRatings,
      photoUrl: apiDoctor.image || 'assets/imageDoctor.png',
      addresses: apiDoctor.addresses.map((addr) => ({
        address: addr.second,
        institution: {
          institutionId: addr.first.institutionId,
          institutionName: addr.first.institutionName,
        },
      })),
      currentAddressIndex: 0,
      schedule: groupSchedulesByDate(apiDoctor.schedules),
    };
  }
}

File: mobile/app/src/main/java/com/medipath/core/services/SearchService.kt
package com.medipath.core.services

import com.medipath.core.models.DoctorCommentsResponse
import com.medipath.core.models.InstitutionDetail
import com.medipath.core.models.DoctorScheduleResponse
import com.medipath.core.models.InstitutionDetailResponse
import com.medipath.core.models.InstitutionDoctorsResponse
import com.medipath.core.models.SearchResponse
import retrofit2.Response
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface SearchService {
    @GET("/api/search/{query}")
    suspend fun search(
        @Path("query") query: String, 
        @Query("type") type: String, 
        @Query("city") city: String? = null, 
        @Query("specialisations") specialisations: String? = null
    ): Response<SearchResponse>
    
    @GET("/api/doctors/{doctorId}/schedules")
    suspend fun getDoctorSchedules(
        @Path("doctorId") doctorId: String,
        @Query("institution") institutionId: String? = null
    ): Response<DoctorScheduleResponse>
    
    @GET("/api/comments/doctor/{id}")
    suspend fun getDoctorComments(
        @Path("id") doctorId: String
    ): Response<DoctorCommentsResponse>

    @GET("/api/institution/{institutionId}/doctors")
    suspend fun getInstitutionDoctors(
        @Path("institutionId") institutionId: String
    ): Response<InstitutionDoctorsResponse>

    @GET("/api/institution/{id}")
    suspend fun getInstitution(
        @Path("id") id: String,
        @Query("fields") fields: List<String>? = null
    ): Response<InstitutionDetailResponse>
}